certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+rule correctDeadlineCalculation() {
+    env e; env e2;
+    require e.block.timestamp > saneTimeConfig();
+    require e2.block.timestamp >= e.block.timestamp;
+
+    // uint64 lastReportRefSlotA; uint64 lastConsensusRefSlotA; uint64 lastConsensusVariantIndexA;
+    // lastReportRefSlotA, lastConsensusRefSlotA, lastConsensusVariantIndexA = helper_getReportingState(e);
+
+    uint256 slot; bytes32 report; uint256 consensusVersion;
+    submitReport(e, slot, report, consensusVersion);
+
+    // uint64 lastReportRefSlotB; uint64 lastConsensusRefSlotB; uint64 lastConsensusVariantIndexB;
+    // lastReportRefSlotB, lastConsensusRefSlotB, lastConsensusVariantIndexB = helper_getReportingState(e2);
+
+    // require lastReportRefSlotB > lastReportRefSlotA;  // the report was submitted successfully
+
+    uint256 refSlot; uint256 reportProcessingDeadlineSlot;
+    refSlot, reportProcessingDeadlineSlot = getCurrentFrame(e2);
+
+    uint256 initialEpoch; uint256 epochsPerFrame; uint256 fastLaneLengthSlots;
+    initialEpoch, epochsPerFrame, fastLaneLengthSlots = getFrameConfig(e2);
+
+    uint256 slotsPerEpoch; uint256 secondsPerSlot; uint256 genesisTime;
+    slotsPerEpoch, secondsPerSlot, genesisTime = getChainConfig(e2);
+
+    uint256 frameSize = epochsPerFrame * slotsPerEpoch;
+    assert (ghostReportHash != 0) => reportProcessingDeadlineSlot == refSlot + frameSize;  // ghostReportHash != 0 ensures report was submitted
+}
+
+
+// 21. submitReport() - the same oracle should not report the same slot+hash twice (cannot double vote for same report)
+// Status: Timeout (no saneTimeConfig)
+// https://prover.certora.com/output/80942/353d54264ae746bd909eeed800613c53/?anonymousKey=665e5f184d4d34fd75bf1f558c43b6f962f16c77
+// Status: - 
+// https://prover.certora.com/output/80942/03e046b378e84977983a9e885db3bf07?anonymousKey=16291f5eddc7374fe1faa5754233b8da62dd71e7
+rule memberCannotDoubleVote() {
+    env e; env e2;
+    require e.block.timestamp > saneTimeConfig();
+    require e2.block.timestamp >= e.block.timestamp;
+
+    uint256 lastProcessingRefSlot = helper_getLastProcessingRefSlot(e);
+
+    uint256 slot; bytes32 report; uint256 consensusVersion;
+    require slot > lastProcessingRefSlot;  // otherwise the report doesn't matter
+
+    submitReport(e, slot, report, consensusVersion);
+    submitReport@withrevert(e2, slot, report, consensusVersion);
+    assert (ghostReportHash != 0) => lastReverted;  // if the first submitReport() was successful then the second should revert
+}
+
+
+// 22. submitReport() - single call to submit report increases only one support for one variant
+//                      sum of supports of all variants < total members,
+//                      because if Oracle member changes its mind, its previous support is removed
+// Status: -
+// 
+
+
+// 23. submitReport() - when new frame starts _reportVariantsLength should reset to 1
```

Comment:
I don't understand the implementation of this rule.
Why doesn't it look like:
frame_after != frame_before => length == 1